src/compiler/sys.ts

```diff
@@ -293,6 +293,50 @@ namespace ts {
                     removeFile: removeFile
                 };
             }
+            
+            function createWatchedFileSet() {
+                let watchedDirectories: { [path: string]: FileWatcher } = {};
+                let watchedFiles: { [fileName: string]: (fileName: string, removed?: boolean) => void; } = {};
+                
+                function addFile(fileName: string, callback: (fileName: string, removed?: boolean) => void): WatchedFile {
+                    const file: WatchedFile = { fileName, callback };
+                    let watchedPaths = Object.keys(watchedDirectories);

```

Comment:
not sure that I understand why do we use map-list structure if ultimately it is still doing linear scan over the array of keys (getting those also incur extra array allocation)